kotlinx-coroutines-core/common/src/flow/internal/FlowScope.kt

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright 2016-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.coroutines.flow.internal
+
+import kotlinx.coroutines.*
+import kotlinx.coroutines.flow.*
+import kotlinx.coroutines.internal.*
+import kotlinx.coroutines.intrinsics.*
+import kotlin.coroutines.*
+import kotlin.coroutines.intrinsics.*
+
+/**
+ * Creates a flow that also provides a [CoroutineScope] for each collector.
+ *
+ * Provided [CoroutineScope] is similar to [coroutineScope] with the only exception that it *ties* lifecycle of children
+ * and itself regarding the cancellation, thus being cancelled when one of the children becomes cancelled.
+ * For example:
+ * ```
+ * scopedFlow {
+ *     launch {
+ *         throw CancellationException()
+ *     }
+ * } // <- CE will be rethrown here
+ * ```
+ *
+ * Basically, it is shorthand for:
+ * ```
+ * flow {
+ *     coroutineScope {
```

Comment:
I'll need a separate `flowScope` primitive, too, for other operators I'm working on, so it makes sense to have a separate internal `flowScope` primitive too and say here that `scopedFlow { ... }` is a shortcut for `flow { flowScope { ... } }` that sames on lambdas (one lambda instead of two)